Added round_with_scale_factor arg to ATen

vfdev-5 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #100079
* #97869
* __->__ #97868

Addresses #62396 following the strategy described in https://github.com/pytorch/pytorch/pull/64983#issuecomment-1026177629.

Fixing output size to match opencv, scikit-image, scipy if scale factor is specified on ATen side only due to JIT FC.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @EikanWang

vfdev-5 committed:
Added round_with_scale_factor arg to ATen

Addresses #62396
Fixing output size to match opencv, scikit-image, scipy if scale factor is specified

[ghstack-poisoned]

vfdev-5 committed:
Update on "Added round_with_scale_factor arg to ATen"


Addresses #62396
Fixing output size to match opencv, scikit-image, scipy if scale factor is specified

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mikaylagawarecki commented on aten/src/ATen/native/UpSample.cpp:
```diff
@@ -24,7 +25,10 @@ TORCH_API c10::SmallVector<int64_t, 3> compute_output_size(
+      const double d = round_with_scale_factor ? 0.5 : 0.0;
+      // if round_with_scale_factor=true we perform round (i.e. int(0.5 + x)) to match opencv, scipy,
+      // scikit-image output size
+      const double odim = d + static_cast<double>(input_size[i+2]) * scale_factors.value()[i];
```
Just for my own understanding, is there a difference/benefit of this over using `std::round` when `round_with_scale_factor=True`?

vfdev-5 replied:
No, there is no benefit other than "more simple" code. Otherwise it would be something like
```c++
double odim = static_cast<double>(input_size[i+2]) * scale_factors.value()[i];
if (round_with_scale_factor) {
    odim = std::round(odim);
}
```

vfdev-5 committed:
Update on "Added round_with_scale_factor arg to ATen"


Addresses #62396 following the strategy described in https://github.com/pytorch/pytorch/pull/64983#issuecomment-1026177629.

Fixing output size to match opencv, scikit-image, scipy if scale factor is specified on ATen side only due to JIT FC.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 EikanWang

[ghstack-poisoned]

vfdev-5 committed:
Update on "Added round_with_scale_factor arg to ATen"


Addresses #62396 following the strategy described in https://github.com/pytorch/pytorch/pull/64983#issuecomment-1026177629.

Fixing output size to match opencv, scikit-image, scipy if scale factor is specified on ATen side only due to JIT FC.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 EikanWang

[ghstack-poisoned]

vfdev-5 commented:
@lezcano in those ops the output size is already computed so there is no need of this new (temporary) argument.

lezcano commented:
Then you shouldn't modify the quantized versions of these functions, as those are only dispatched from
```
- func: upsample_bilinear2d(Tensor self, SymInt[2] output_size, bool align_corners, float? scales_h=None, float? scales_w=None) -> Tensor
```
not from the `_vec` variants.

vfdev-5 committed:
Update on "Added round_with_scale_factor arg to ATen"


Addresses #62396 following the strategy described in https://github.com/pytorch/pytorch/pull/64983#issuecomment-1026177629.

Fixing output size to match opencv, scikit-image, scipy if scale factor is specified on ATen side only due to JIT FC.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 EikanWang

[ghstack-poisoned]

vfdev-5 committed:
Update on "Added round_with_scale_factor arg to ATen"


Addresses #62396 following the strategy described in https://github.com/pytorch/pytorch/pull/64983#issuecomment-1026177629.

Fixing output size to match opencv, scikit-image, scipy if scale factor is specified on ATen side only due to JIT FC.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 EikanWang

[ghstack-poisoned]

vfdev-5 commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -632,15 +634,15 @@ at::Tensor interpolate(
         _output_size(input, 2, size, scale_factors),
-        scale_factors_1,
-        scale_factors_2);
+        c10::make_optional(scale_factors_1),
```
I need to explicitly put scale factor as c10::optional otherwise composite [`at::upsample_nearest2d`](https://github.com/pytorch/pytorch/pull/97868/files#diff-5235eb89d3bbcfdbf84c8ad0f96b2c309f3ccb542501b26946a0a47becb7d3eaR150) is called with double scale factor interpreted as bool for `round_with_scale_factor`.

lezcano reviewed (approved):
What a pain of a PR this must have been.

In https://github.com/pytorch/pytorch/pull/85638, these functions were made composite implicit. In theory, it should be possible for most of the subsystems that this PR touches to assume that these functions are CompositeImplicit and delete the special handling logic they have. Now, this would be a pain to land from OSS, so we're leaving things as they are for now.

mikaylagawarecki commented on torch/backends/_nnapi/serializer.py:
```diff
@@ -1549,19 +1549,17 @@ def add_adaptive_avg_pool2d(self, node):
-            scale_h_ctype, scale_h_arg = self.get_constant_value(scale_h_jit)
-            scale_w_ctype, scale_w_arg = self.get_constant_value(scale_w_jit)
+            scale_h_ctype, scale_h_arg = self.get_constant_value(arg2_jit)
+            scale_w_ctype, scale_w_arg = self.get_constant_value(arg3_jit)
```
Just to check, does the new rounding logic need to be added [here in this function](https://github.com/pytorch/pytorch/blob/7c3b66d277ff58250d6fdb56ca4480df4cb6a30a/torch/backends/_nnapi/serializer.py#L1606-L1607) as well? I think python `int()` will round towards zero

mikaylagawarecki reviewed (commented):
Thanks @vfdev-5, this looks good to me! 


edit: Was looking at the PR stacked above this which adds testing and noticed the decomp related failures. These are because the corresponding decompositions for these functions have not been updated. 

For example, there is a python equivalent of `compute_output_size` used for the decompositions [here](https://github.com/pytorch/pytorch/blob/main/torch/_decomp/decompositions.py#L2003) and a decomposition for [upsample_nearest_2d](https://github.com/pytorch/pytorch/blob/main/torch/_decomp/decompositions.py#L2050)

We need to update the decompositions as well, as these are used by `torch.compile` and not updating them would lead to divergences between the `torch.compile`-ed version of upsample and eager mode.

vfdev-5 committed:
Update on "Added round_with_scale_factor arg to ATen"


Addresses #62396 following the strategy described in https://github.com/pytorch/pytorch/pull/64983#issuecomment-1026177629.

Fixing output size to match opencv, scikit-image, scipy if scale factor is specified on ATen side only due to JIT FC.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 EikanWang

[ghstack-poisoned]

vfdev-5 committed:
Update on "Added round_with_scale_factor arg to ATen"


Addresses #62396 following the strategy described in https://github.com/pytorch/pytorch/pull/64983#issuecomment-1026177629.

Fixing output size to match opencv, scikit-image, scipy if scale factor is specified on ATen side only due to JIT FC.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 EikanWang

[ghstack-poisoned]

vfdev-5 committed:
Update on "Added round_with_scale_factor arg to ATen"


Addresses #62396 following the strategy described in https://github.com/pytorch/pytorch/pull/64983#issuecomment-1026177629.

Fixing output size to match opencv, scikit-image, scipy if scale factor is specified on ATen side only due to JIT FC.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 EikanWang

[ghstack-poisoned]

mikaylagawarecki reviewed (approved):
LGTM, thanks!

vfdev-5 commented:
@pytorchbot merge

osalpekar commented:
@pytorchbot revert -m "Caused breakages in the glow compiler - see [D45374622](https://www.internalfb.com/diff/D45374622) for more details" -c ghfirst

osalpekar commented:
cc @mikaylagawarecki for help with the internal breakages
